api_app/connectors_manager/connectors/opencti.py

```diff
@@ -0,0 +1,199 @@
+from pycti.api.opencti_api_client import File
+import pycti
+
+from tests.mock_utils import patch, if_mock_connections
+from api_app import helpers
+from api_app.connectors_manager import classes
+
+
+INTELOWL_OPENCTI_TYPE_MAP = {
+    "ip": {
+        "v4": "ipv4-addr",
+        "v6": "ipv6-addr",
+    },
+    "domain": "domain-name",
+    "url": "url",
+    # type hash is combined with file
+    "generic": "x-opencti-text",  # misc field, so keeping text
+    "file": "file",  # hashes: md5, sha-1, sha-256
+}
+
+
+class OpenCTI(classes.Connector):
+    def set_params(self, params):
+        self.ssl_verify = params.get("ssl_verify", True)
+        self.tlp = params.get(
+            "tlp", {"type": "white", "color": "#FFFFFF", "x_opencti_order": 0}
+        )
+        self.proxies = params.get("proxies", {})
+        self.__url_name = self._secrets["url_key_name"]
+        self.__api_key = self._secrets["api_key_name"]
+
+    def get_observable_type(self) -> str:
+        if self._job.is_sample:
+            obs_type = INTELOWL_OPENCTI_TYPE_MAP["file"]
+        elif self._job.observable_classification == "hash":
+            matched_hash_type = helpers.get_hash_type(self._job.observable_name)
+            if matched_hash_type in [
+                "md5",
+                "sha-1",
+                "sha-256",
+            ]:  # sha-512 not supported
+                obs_type = INTELOWL_OPENCTI_TYPE_MAP["file"]
+            else:
+                obs_type = INTELOWL_OPENCTI_TYPE_MAP["generic"]  # text
+        elif self._job.observable_classification == "ip":
+            ip_version = helpers.get_ip_version(self._job.observable_name)
+            if ip_version == 4 or ip_version == 6:
+                obs_type = INTELOWL_OPENCTI_TYPE_MAP["ip"][f"v{ip_version}"]  # v4/v6
+            else:
+                obs_type = INTELOWL_OPENCTI_TYPE_MAP["generic"]  # text
+        else:
+            obs_type = INTELOWL_OPENCTI_TYPE_MAP[self._job.observable_classification]
+
+        return obs_type
+
+    def generate_observable_data(self) -> dict:
+        observable_data = {"type": self.get_observable_type()}
+        if self._job.is_sample:
+            observable_data["name"] = self._job.file_name
+            observable_data["hashes"] = helpers.generate_hashes(self.job_id)
+        elif (
+            self._job.observable_classification == "hash"
+            and observable_data["type"] == "file"
+        ):
+            # add hash instead of value
+            matched_type = helpers.get_hash_type(self._job.observable_name)
+            observable_data["hashes"] = {matched_type: self._job.observable_name}
+        else:
+            observable_data["value"] = self._job.observable_name
+
+        return observable_data
+
+    @property
+    @cache_memoize(timeout=sys.maxsize)
+    def organization_id(self) -> str:  # called only once (cached indefinitely)
+        # Create author (if not exists)
+        org = pycti.Identity(self.opencti_instance).create(
```

Comment:
If you see the PyCTI docs,

> The Threat-Actor entity will only be created if it doesn’t exists By setting update to True it acts like an upsert and updates fields of an existing Threat-Actor entity.

The same applies to the `pycti.Identity.create` method (they do not document this but I looked at the [code](https://github.com/OpenCTI-Platform/client-python/blob/f75785e4cd6c249eb1db3dc53126cc0ca98370bb/pycti/entities/opencti_identity.py#L8)). So we don't need to have our own caching, we can just provide `update=True` keyword argument.

Reply:
A way to test if this does work or not would be to run 2 connector calls and see that only one Identity was created.

Reply:
If you were concerned that it would create two identities, it wouldn't. I thought we were avoiding making that call after first one.

Reply:
I agree that we do not need our own caching and handling, let's leave it to the API.